docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,159 @@ the worker pool has a finite size and even with the full batching algorithm in t
+|-----:|--------:|--------:|-----------------------------------------------------------------------:|
+| 0x6d |    0x00 |      -1 | 0x7b2273657175656e6365223a31333030372c22626f756e64223a353031333030377d |
+| 0x6d |  0xa711 |      -1 |                                     0x7b2276657...3334227d (216 bytes) |
+| 0x6d |  0xb2cf |      -1 |                                     0x7b2276657...3930227d (380 bytes) |
+
+The single ``key`` reflects that these values are all part of the same sequence of values. The column IDs are
+``VAR_LONG`` encoded; ``0x00`` maps to zero, and the associated value decodes to a ``SequenceAndBound``, in this
+example ``{"sequence":13007,"bound":5013007}``. This means that the value at the column associated with sequence
+``13007`` (which encodes to ``b2cf``) is the currently agreed value; the value in column ``a711`` is not current.
+
+The main reason for this scheme is that while the validity bound itself changes frequently (approximately every
+``ADVANCEMENT_QUANTUM`` timestamps, which is 5,000,000 at time of writing), the value itself changes much less often.
+For example, with transactions2, in the absence of rollbacks, the value is expected to change just once in the
+service's lifetime, keeping track of the point where reading from transactions2 should begin.
+To further support this scheme, the coordination service recognises identity transformations from the current value,
+and doesn't write a new value in these cases.
+
+We have not implemented cleanup of the coordination service at this time, as we don't expect the number of values
+in a coordination sequence to be large. Furthermore, we don't ever range scan this table.
+
+#### Determining Transaction Schema Versions
+
+We still need to retain the ability to read from the old transactions table, when checking if/when values written
+before the transactions2 migration occurred. Thus, we replace the old ``TransactionService`` (that only reads/writes
+from transactions1) with one that delegates between multiple TransactionServices (concretely, just V1 and V2 at this
+time) depending on the specific timestamps involved in calls.
+
+The mechanism for determining which ``TransactionService`` to use goes through the aforementioned
+``CoordinationService``. We agree on a ``TimestampPartitioningMap``, which is a mapping of timestamps to transaction
+schema versions. This is a range map, as we want to support rollbacks easily. The map also is guaranteed to span the
+ranges ``[1, +∞)`` and be connected; note that this does not mean future behaviour is fixed, since in practice this map
+is read together with a validity bound. Thus, even though the map will contain as a key some range ``[C, +∞)`` for a
+constant ``C``, it should be valid up to some point ``V > C`` - and behaviour at timestamps after ``V`` may subsequently
+be changed.
+
+Note that the ranges are based on the start timestamp. To give a concrete example, if our range map is
```

Comment:
Think this is a bit of a leap. We've got to the point where we've cut up timestamp space into difference ranges. But the connecting bit (in my mind) is:
> Since we accessed the original `TransactionService` via the `startTimestamp`, we will use the `startTimestamp` against the range map to find out the behaviour at that point.

Or something along those lines.

Reply:
Yep, added pretty much verbatim (apart from not using start timestamp as a symbol)